drivers/virtio/virtio_pci.c

```diff
@@ -0,0 +1,688 @@
+	 */
+	data->common_cfg->device_status = 0;
+
+	while (data->common_cfg->device_status != 0) {
+	}
+}
+
+static void virtio_pci_notify_queue(const struct device *dev, uint16_t queue_idx)
+{
+	struct virtio_pci_data *data = dev->data;
+
+	k_spinlock_key_t key = k_spin_lock(&data->notify_lock);
+
+	data->common_cfg->queue_select = queue_idx;
+	barrier_dmem_fence_full();
+	/*
+	 * Because currently we are not negotiating VIRTIO_F_NOTIFICATION_DATA
+	 * and VIRTIO_F_NOTIF_CONFIG_DATA, in order to notify the queue we have
+	 * to write its index to notify_cfg at the offset
+	 * cap.offset + queue_notify_off * notify_off_multiplier,
+	 * which here is reduced to queue_notify_off * notify_off_multiplier,
+	 * because data->notify_cfg was mapped in virtio_pci_map_cap() to start
+	 * at cap.offset. See spec 4.1.4.4 for the offset formula, and spec
+	 * 4.1.5.2 and spec 4.1.5.2.1 for the value written
+	 */
+	size_t notify_off = data->common_cfg->queue_notify_off * data->notify_off_multiplier;
+	volatile uint16_t *notify_addr = (uint16_t *)(data->notify_cfg + notify_off);
+
+	*notify_addr = sys_cpu_to_le16(queue_idx);
+
+	k_spin_unlock(&data->notify_lock, key);
+}
+
+/*
+ * According to the spec 4.1.3.1, PCI virtio driver must use n byte accesses for n byte fields,
+ * except for 64 bit fields where 32 bit accesses have to be used, so we are using this
+ * function to write 64 bit values to 64 bit fields
+ */
+static void virtio_pci_write64(uint64_t val, uint64_t *dst)
+{
+	uint64_t val_le = sys_cpu_to_le64(val);
+
+	((uint32_t *)dst)[0] = val_le & GENMASK64(31, 0);
+	((uint32_t *)dst)[1] = (val_le & GENMASK64(63, 32)) >> 32;
+}
+
+static int virtio_pci_set_virtqueue(
+	const struct device *dev, uint16_t virtqueue_n, struct virtq *virtqueue)
+{
+	struct virtio_pci_data *data = dev->data;
+
+	data->common_cfg->queue_select = virtqueue_n;
+	barrier_dmem_fence_full();
+
+	uint16_t max_queue_size = sys_le16_to_cpu(data->common_cfg->queue_size);
+
+	if (max_queue_size < virtqueue->num) {
+		LOG_ERR(
+			"virtio pci device doesn't support queue %d bigger than %d, tried to set one with size %d",
+			virtqueue_n,
+			max_queue_size,
+			virtqueue->num
+		);
+		return -EINVAL;
+	}
+	data->common_cfg->queue_size = sys_cpu_to_le16(virtqueue->num);
+	virtio_pci_write64(
+		k_mem_phys_addr(virtqueue->desc), (void *)&data->common_cfg->queue_desc
+	);
+	virtio_pci_write64(
+		k_mem_phys_addr(virtqueue->avail), (void *)&data->common_cfg->queue_driver
+	);
+	virtio_pci_write64(
+		k_mem_phys_addr(virtqueue->used), (void *)&data->common_cfg->queue_device
+	);
+	data->common_cfg->queue_msix_vector = sys_cpu_to_le16(VIRTIO_PCI_MSIX_NO_VECTOR);
+	data->common_cfg->queue_enable = sys_cpu_to_le16(1);
+
+	return 0;
+}
+
+static int virtio_pci_set_virtqueues(
+	const struct device *dev, virtio_enumerate_queues cb, void *opaque)
+{
+	struct virtio_pci_data *data = dev->data;
+	uint16_t queue_count = sys_le16_to_cpu(data->common_cfg->num_queues);
```

Comment:
As feedback from the result of implementing virtio-mmio in https://github.com/zephyrproject-rtos/zephyr/pull/89460.

The virtio-mmio does not have a register equivalent to `num_queues`, so it is not possible to determine the upper limit of the number of queues.

In the Linux API `find_vqs`, the number of queues is explicitly specified from the driver side.

In the same way, it is recommended that an argument be added to the virtio_init_virtqueues API.